android/cli/lib/gradle-wrapper.js

```diff
@@ -0,0 +1,382 @@
+				const index = stringBuffer.lastIndexOf('\n');
+				if (index <= 0) {
+					return;
+				}
+
+				// Log the received messages, split by newline. (Strip out carriage returns on Windows.)
+				const messageArray = stringBuffer.substr(0, index).split('\n');
+				for (const nextMessage of messageArray) {
+					logFunction('[GRADLE] ' + nextMessage.replace(/\r/g, ''));
+				}
+
+				// Remove the above logged strings from the buffer.
+				// This keeps the trailing characters that don't end with a newline (yet).
+				if ((index + 1) < stringBuffer.length) {
+					stringBuffer = stringBuffer.substr(index + 1);
+				} else {
+					stringBuffer = '';
+				}
+			};
+		};
+
+		// Run the gradlew command line async.
+		return new Promise((resolve, reject) => {
+			const childProcess = exec(commandLineString, { cwd: this._gradlewDirPath });
+			if (this._logger) {
+				childProcess.stdout.on('data', createReadableDataHandlerUsing(this._logger.info));
+				childProcess.stderr.on('data', createReadableDataHandlerUsing(this._logger.error));
+			}
+			childProcess.on('error', reject);
+			childProcess.on('exit', (exitCode) => {
+				if (exitCode === 0) {
+					resolve();
+				} else {
+					reject(`"gradlew" tool returned exit code: ${exitCode}`);
+				}
+			});
+		});
+	}
+
+	/**
+	 * Copies the "gradlew" file tree from the given "templateDirPath" to this object's assigned "directoryPath".
+	 * This must be done before calling this object's clean(), assembleRelease(), assembleDebug(), or run() methods
+	 * since the "gradlew" scripts and libraries must be copied to the root project's directory.
+	 * @param {String} templateDirPath
+	 * Path to a directory tree containing the "gradlew" scripts and libraries. Cannot be null/undefined.
+	 *
+	 * Expected to be set to: "<RootFolder>/Titanium/mobilesdk/<Platform>/<Version>/android/templates/gradle"
+	 */
+	async installTemplate(templateDirPath) {
+		// Validate argument.
+		if (typeof templateDirPath !== 'string') {
+			throw new Error('Argument "templateDirPath" must be set to a string.');
+		}
+
+		// Copy the gradle template directory tree to the directory this object references.
+		// Note: The copy function does not copy file permissions. So, we must re-add execute permissions.
+		//       0o755 = User Read/Write/Exec, Group Read/Execute, Others Read/Execute
+		const destinationDirPath = this._gradlewDirPath;
+		await fs.copy(templateDirPath, destinationDirPath);
+		await fs.chmod(path.join(destinationDirPath, 'gradlew'), 0o755);
+		await fs.chmod(path.join(destinationDirPath, 'gradlew.bat'), 0o755);
+
+		// Fetch proxy server information, if configured.
+		let proxyUrl = null;
+		await new Promise((resolve) => {
+			const runOptions = { shell: true, windowsHide: true };
+			appc.subprocess.run('appc', [ '-q', 'config', 'get', 'proxyServer' ], runOptions, (exitCode, out) => {
+				try {
+					if (!exitCode && out && (out.length > 0)) {
+						proxyUrl = url.parse(out.trim());
```

Comment:
<!--
  0 failure: 
  1 warning:  android/cli/lib/g...
  
  
  DangerID: danger-id-Danger;
  File: android/cli/lib/gradle-wrapper.js;
  Line: 214;
-->

- :warning: android/cli/lib/gradle-wrapper.js line 214 – 'url.parse' was deprecated since v11.0.0. Use 'url.URL' constructor instead. (node/no-deprecated-api)